pymc_marketing/clv/models/pareto_nbd.py

```diff
@@ -0,0 +1,537 @@
+            - (r + x) * log(alpha + T)
+            - s * log(beta + T)
+        )
+        second_term = log(r + x) + log(beta + T) - log(alpha + T)
+        third_term = log(
+            (1 - ((beta + T) / (beta + T + future_t)) ** (s - 1)) / (s - 1)
+        )
+
+        exp_purchases = exp(first_term + second_term + third_term - loglike)
+
+        return exp_purchases.transpose(
+            "chain", "draw", "customer_id", missing_dims="ignore"
+        )
+
+    def expected_purchases_new_customer(
+        self,
+        t: Union[np.ndarray, pd.Series],
+    ) -> xarray.DataArray:
+        r"""
+            Expected number of purchases for a new customer across :math:`t` time periods. See
+            equation (27) of [2]_.
+
+            http://brucehardie.com/notes/009/pareto_nbd_derivations_2005-11-05.pdf
+
+            Adapted from lifetimes package
+        https://github.com/CamDavidsonPilon/lifetimes/blob/41e394923ad72b17b5da93e88cfabab43f51abe2/lifetimes/fitters/pareto_nbd_fitter.py#L359
+
+            Parameters
+            ----------
+            t: array_like
+                Number of time periods over which to estimate purchases.
+        """
+        r, alpha, s, beta = self._unload_params()
+
+        t = np.asarray(t)
+
+        r, alpha, s, beta = self._unload_params()  # type: ignore [has-type]
+        first_term = r * beta / alpha / (s - 1)
+        second_term = 1 - (beta / (beta + t)) ** (s - 1)
+
+        return (first_term * second_term).transpose(
+            "chain", "draw", "t", missing_dims="ignore"
+        )
+
+    def expected_probability_alive(
+        self,
+        future_t: Union[int, float] = 0,
+        customer_id: Union[np.ndarray, pd.Series] = None,
+        recency: Union[np.ndarray, pd.Series] = None,
+        frequency: Union[np.ndarray, pd.Series] = None,
+        T: Union[np.ndarray, pd.Series] = None,
+    ) -> xarray.DataArray:
+        r"""
+        Compute the probability that a customer with history :math:`frequency`, :math:`recency`, and :math:`T`
+        is currently alive. Can also estimate alive probability :math:`future_t` periods in the future.
+
+        If no customer data is provided, probabilities for all customers in model fit dataset are returned.
+
+        See equation (18) from [3]_.
+
+        Parameters
+        ----------
+        future_t: scalar
+            Number of time periods in the future to estimate alive probability; defaults to 0.
+        customer_id: array_like
+            Customer labels.
+        recency: array_like
+            Number of time periods between the customer's first and most recent purchases.
+        frequency: array_like
+            Number of repeat purchases per customer.
+        T: array_like
+            Number of time periods since the customer's first purchase.
+            Model assumptions require T >= recency.
+        """
+        x, t_x, T = self._process_customers(customer_id, frequency, recency, T)
```

Comment:
As before... can the inputs of `self._process_customers` be `None`? We need to have consistent types ;)